src/python/pants/testutil/rule_runner.py

```diff
@@ -0,0 +1,238 @@
+from pants.engine.fs import PathGlobs, PathGlobsAndRoot, Snapshot, Workspace
+from pants.engine.goal import Goal
+from pants.engine.internals.native import Native
+from pants.engine.internals.scheduler import SchedulerSession
+from pants.engine.internals.selectors import Params
+from pants.engine.process import InteractiveRunner
+from pants.engine.rules import QueryRule, Rule
+from pants.engine.target import Target, WrappedTarget
+from pants.init.engine_initializer import EngineInitializer
+from pants.option.global_options import ExecutionOptions, GlobalOptions
+from pants.option.options_bootstrapper import OptionsBootstrapper
+from pants.source import source_root
+from pants.testutil.option_util import create_options_bootstrapper
+from pants.util.collections import assert_single_element
+from pants.util.contextutil import temporary_dir
+from pants.util.dirutil import recursive_dirname, safe_file_dump, safe_mkdir, safe_open
+from pants.util.meta import frozen_after_init
+from pants.util.ordered_set import FrozenOrderedSet
+
+_P = TypeVar("_P")
+
+
+@dataclass(frozen=True)
+class GoalRuleResult:
+    exit_code: int
+    stdout: str
+    stderr: str
+
+    @staticmethod
+    def noop() -> "GoalRuleResult":
+        return GoalRuleResult(0, stdout="", stderr="")
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class RuleRunner:
+    build_root: str
+    build_config: BuildConfiguration
+    scheduler: SchedulerSession
+
+    def __init__(
+        self,
+        *,
+        rules: Optional[Iterable] = None,
+        target_types: Optional[Iterable[Type[Target]]] = None,
+        objects: Optional[Dict[str, Any]] = None,
+        context_aware_object_factories: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self.build_root = os.path.realpath(mkdtemp(suffix="_BUILD_ROOT"))
+        safe_mkdir(self.build_root, clean=True)
+        safe_mkdir(self.pants_workdir)
+        BuildRoot().path = self.build_root
+
+        # TODO: Redesign rule registration for tests to be more ergonomic and to make this less
+        #  special-cased.
+        all_rules = (
+            *(rules or ()),
+            *source_root.rules(),
+            QueryRule(WrappedTarget, (Address, OptionsBootstrapper)),
+        )
+        build_config_builder = BuildConfiguration.Builder()
+        build_config_builder.register_aliases(
+            BuildFileAliases(
+                objects=objects, context_aware_object_factories=context_aware_object_factories
+            )
+        )
+        build_config_builder.register_rules(all_rules)
+        build_config_builder.register_target_types(target_types or ())
+        self.build_config = build_config_builder.create()
+
+        options_bootstrapper = OptionsBootstrapper.create(
+            env={}, args=["--pants-config-files=[]"], allow_pantsrc=False
+        )
+        global_options = options_bootstrapper.bootstrap_options.for_global_scope()
+        local_store_dir = global_options.local_store_dir
```

Comment:
We need to still emulate `isolated_local_store`, which is used in `fs_test.py` to avoid using the global `lmdb_store`. Saving for a followup.